client/castApp/src/castinit.c

```diff
@@ -65,13 +67,124 @@ static void casthook(initHookState state)
     epicsAtExit(&castexit, NULL);
 }
 
+/*
+    Example call: addReccasterEnvVars("SECTOR") or addReccasterEnvVars("SECTOR", "BUILDING")
+    Appends the given env variables to the extra_envs list to be sent in addition to the default_envs array
+*/
+static void addReccasterEnvVars(caster_t* self, int argc, char **argv)
+{
+    size_t i, j;
+    int ret = 0;
+
+    argv++; argc--; /* skip function arg */
+    if(argc < 1) {
+        errlogSevPrintf(errlogMinor, "At least one argument expected for addReccasterEnvVars\n");
+        return;
+    }
+
+    epicsMutexMustLock(self->lock);
+    if(self->shutdown) {
+        /* shutdown in progress, silent no-op */
+        epicsMutexUnlock(self->lock);
+        return;
+    }
+    else if(self->current != casterStateInit) {
+        /* Attempt to add after iocInit(), when we may be connected.
+           To fully support, would need to force reconnect or resend w/ updated envs list. */
+        errlogSevPrintf(errlogMinor, "addReccasterEnvVars called after iocInit() when reccaster might already be connected. Not supported\n");
+        epicsMutexUnlock(self->lock);
+        return;
+    }
+    int new_extra_envs_size = self->num_extra_envs + argc;
+    int num_new_extra_envs = self->num_extra_envs;
+
+    char **new_extra_envs = calloc(new_extra_envs_size, sizeof(*new_extra_envs));
+    if(new_extra_envs == NULL) {
+        errlogSevPrintf(errlogMajor, "Error in memory allocation of new_extra_envs from addReccasterEnvVars\n");
+        epicsMutexUnlock(self->lock);
+        return;
+    }
+    /* copy self->extra_envs into new_extra_envs with room for new envs */
+    for(i=0; i < self->num_extra_envs; i++) {
+        if((new_extra_envs[i] = strdup(self->extra_envs[i])) == NULL) {
+            errlogSevPrintf(errlogMinor, "strdup error from addReccasterEnvVars\n", new_extra_envs[j]);
+            ret = 1;
+            break;
+        }
+    }
+    int found_dup;
+    /* sanitize input - check for dups and empty args */
+    if(!ret) {
+        for(i=0; i < argc; i++) {
+            if(argv[i] == NULL) {
+                errlogSevPrintf(errlogMinor, "Arg is NULL for addReccasterEnvVars\n");
+                continue;
+            }
+            else if(argv[i][0] == '\0') {
+                errlogSevPrintf(errlogMinor, "Arg is empty for addReccasterEnvVars\n");
+                continue;
+            }
+            found_dup = 0;
+            /* check if dup in self->extra_envs. doesn't check if arg is in default_envs right now */
+            for(j = 0; j < num_new_extra_envs && new_extra_envs[j]; j++) {
```

Comment:
I think that this could be split up into a test within the for loop. If somehow or other a `NULL` sneaks its way into `self->extra_envs`, then it would no longer check anything past that point.

(I think that the code is also a bit easier to follow with a separate condition in the loop instead of including it as the loop condition)